akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.pattern
+
+import akka.actor._
+import akka.actor.OneForOneStrategy
+import akka.actor.SupervisorStrategy._
+import scala.concurrent.duration._
+
+private object BackoffOnRestartSupervisor {
+  private case class ScheduleRestart(childRef: ActorRef) extends DeadLetterSuppression
+  private case object StartChild extends DeadLetterSuppression
+  private case class ResetRestartCount(lastNumRestarts: Int) extends DeadLetterSuppression
+}
+
+private class BackoffOnRestartSupervisor(
+  childProps: Props,
+  childName: String,
+  minBackoff: FiniteDuration,
+  maxBackoff: FiniteDuration,
+  reset: BackoffReset,
+  randomFactor: Double,
+  strategy: OneForOneStrategy)
+  extends Actor
+  with ActorLogging {
+
+  import BackoffOnRestartSupervisor._
+  import context._
+
+  private var child: Option[ActorRef] = None
+
+  override val supervisorStrategy = OneForOneStrategy(strategy.maxNrOfRetries, strategy.withinTimeRange, strategy.loggingEnabled) {
+    case ex ⇒
+      val defaultDirective: Directive =
+        super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) ⇒ Escalate)
+
+      strategy.decider.applyOrElse(ex, (_: Any) ⇒ defaultDirective) match {
+
+        // Whatever the final Directive is, we will translate all Restarts
+        // to our own Restarts, which involves stopping the child.
+        // directive match {
+        case Restart ⇒
+          val childRef = sender()
+          become({
+            case Terminated(`childRef`) ⇒
+              unbecome()
+              self ! ScheduleRestart(childRef)
+            case _ ⇒
+          }, discardOld = false)
+          Stop
+        case other ⇒ other
+      }
+  }
+
+  // Initialize by starting up and watching the child
+  self ! StartChild
+
+  def receive = waitingToStart(-1, false)
+
+  def waitingToStart(numRestarts: Int, scheduleCounterReset: Boolean): Receive = {
+    case StartChild ⇒
+      val childRef = actorOf(childProps, childName)
+      child = Some(childRef)
+      watch(childRef)
+      if (scheduleCounterReset) {
+        system.scheduler.scheduleOnce(minBackoff, self, ResetRestartCount(numRestarts + 1))
+      }
+      become(watching(childRef, numRestarts + 1))
+    case _ ⇒
+  }
+
+  // Steady state
+  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
+    case ScheduleRestart(`childRef`) ⇒
+      val delay = akka.pattern.BackoffSupervisor.calculateDelay(
+        numRestarts, minBackoff, maxBackoff, randomFactor)
+      system.scheduler.scheduleOnce(delay, self, StartChild)
+      become(waitingToStart(numRestarts, true))
```

Comment:
For the case that reset is ManualReset, do we still actually want to schedule the automatic counter reset?